Use more smart pointers in PDF extension

basuke opened the pull request:
#### 7c456d3a318b50f5eb06050e4a38a1b6cf982dc1
<pre>
Use more smart pointers in PDF extension
<a href="https://bugs.webkit.org/show_bug.cgi?id=279472">https://bugs.webkit.org/show_bug.cgi?id=279472</a>

Reviewed by Chris Dumez and Ryosuke Niwa.

Use smart pointer to protect arguments to be gone.

* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
(WebKit::ByteRangeRequest::protectedStreamLoader):
(WebKit::PDFIncrementalLoader::receivedNonLinearizedPDFSentinel):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidReceiveData):
(WebKit::PDFIncrementalLoader::dataProviderGetBytesAtPosition):
(WebKit::PDFIncrementalLoader::dataProviderGetByteRanges):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::generatePreviewImageForPage):
(WebKit::AsyncPDFRenderer::serviceRequestQueue):
(WebKit::AsyncPDFRenderer::paintTilesForPage):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
(WebKit::PDFDiscretePresentationController::startTransitionAnimation):
(WebKit::PDFDiscretePresentationController::updateLayersForTransitionState):
(WebKit::PDFDiscretePresentationController::deviceOrPageScaleFactorChanged):
(WebKit::PDFDiscretePresentationController::buildRows):
(WebKit::PDFDiscretePresentationController::updateLayersOnLayoutChange):
(WebKit::PDFDiscretePresentationController::updateLayersAfterChangeInVisibleRow):
(WebKit::PDFDiscretePresentationController::repaintForIncrementalLoad):
(WebKit::PDFDiscretePresentationController::RowData::leftPageBackgroundLayer const):
(WebKit::PDFDiscretePresentationController::RowData::rightPageBackgroundLayer const):
(WebKit::PDFDiscretePresentationController::RowData::backgroundLayerForPageIndex const):

Canonical link: <a href="https://commits.webkit.org/283526@main">https://commits.webkit.org/283526@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3952c8f36e9a4592e1d74e9424b127941d4c1e33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66566 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19187 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70600 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17699 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53740 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17459 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11934 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69633 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/42301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57595 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34008 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/38972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16053 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/60884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15327 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72302 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10523 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10555 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61001 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14689 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8657 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/42825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/44008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42568 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -1069,19 +1079,20 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
     auto parentRowLayers = [](RowData& row) {
         ASSERT(row.containerLayer->children().isEmpty());
-        row.containerLayer->addChild(*row.leftPageContainerLayer);
+        row.containerLayer->addChild(row.leftPageContainerLayer.releaseNonNull());
```
This will clear `row.leftPageContainerLayer`. Is that what we want to do here?

cdumez replied:
This looks like a potential bug to me.

aprotyas replied:
@basuke the `releaseNonNull()` calls in this lambda will cause issues because we don't want to clear the `[left|right]PageContainerLayer` members.

You can safely keep dereferencing here.

basuke replied:
Got it. Sorry I read the code wrong.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -548,9 +548,11 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
-        animatingRow.containerLayer->setPosition(layerPosition);
-        animatingRow.containerLayer->setOpacity(1);
-        stationaryRow.containerLayer->setOpacity(1);
+        Ref animatingRowContainerLayer = *animatingRow.containerLayer.get();
```
Why `.get()`?

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -548,9 +548,11 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
+        Ref animatingRowContainerLayer = *animatingRow.containerLayer.get();
+        animatingRowContainerLayer->setPosition(layerPosition);
+        animatingRowContainerLayer->setOpacity(1);
+        Ref stationaryRowContainerLayer = *stationaryRow.containerLayer.get();
```
ditto.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -812,10 +814,11 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
-            row.containerLayer->setOpacity(1);
-            row.containerLayer->removeAnimation("move"_s, { });
-            row.containerLayer->removeAnimation("fade"_s, { });
+            Ref rowContainerLayer = *row.containerLayer.get();
```
Why `.get()`

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -843,22 +846,26 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
             // Previous page pulls down up from the top or right.
-            topLayerRow.containerLayer->setPosition(layerPosition);
-            topLayerRow.containerLayer->setOpacity(layerOpacities[topLayerIndex][startIndex]);
+            Ref topLayerRowContainerLayer = *topLayerRow.containerLayer.get();
```
`.get()` seems useless?

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -843,22 +846,26 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
             if (additionalVisibleRowIndex) {
                 auto& bottomRow = m_rows[topLayerRowIndex + 1];
-                bottomRow.containerLayer->setOpacity(layerOpacities[bottomLayerIndex][startIndex]);
+                Ref bottomRowContainerLayer = *bottomRow.containerLayer.get();
```
`.get()` seems useless?

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -1095,24 +1106,26 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
-        row.contentsLayer->setAnchorPoint({ });
-        row.contentsLayer->setDrawsContent(true);
-        row.contentsLayer->setAcceleratesDrawing(m_plugin->canPaintSelectionIntoOwnedLayer());
+        auto rowContentsLayer = createGraphicsLayer(makeString("Row contents "_s, rowIndex), GraphicsLayer::Type::TiledBacking);
```
`auto` -> `Ref` ? We try not to hide smart pointers behind `auto` anymore.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -1095,24 +1106,26 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
-        row.selectionLayer->setAcceleratesDrawing(true);
-        row.selectionLayer->setBlendMode(BlendMode::Multiply);
-        m_layerIDToRowIndexMap.set(*row.selectionLayer->primaryLayerID(), rowIndex);
+        auto rowSelectionLayer = createGraphicsLayer(makeString("Row selection "_s, rowIndex), GraphicsLayer::Type::TiledBacking);
```
auto -> Ref ?

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -1190,20 +1203,23 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
 
-    m_rowsContainerLayer->setTransform(transform);
+    Ref rowsContainerLayer = *m_rowsContainerLayer.get();
+    rowsContainerLayer->setTransform(transform);
```
Normally, we add a `protectedRowsContainerLayer()` which returns a Ref / RefPtr and call `protectedRowsContainerLayer()->setTransform(transform);`. There is no reason to introduce a local variable for a single function call.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -1572,24 +1594,25 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
 
 RefPtr<GraphicsLayer> PDFDiscretePresentationController::RowData::leftPageBackgroundLayer() const
 {
-    return PDFPresentationController::pageBackgroundLayerForPageContainerLayer(*leftPageContainerLayer);
```
`return PDFPresentationController::pageBackgroundLayerForPageContainerLayer(Ref { *leftPageContainerLayer });`

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
```diff
@@ -145,53 +145,65 @@
-        m_pageBackgroundsContainerLayer = createGraphicsLayer("Page backgrounds"_s, GraphicsLayer::Type::Normal);
-        m_pageBackgroundsContainerLayer->setAnchorPoint({ });
-        scrolledContentsLayer.addChild(*m_pageBackgroundsContainerLayer);
+        auto layer = createGraphicsLayer("Page backgrounds"_s, GraphicsLayer::Type::Normal);
```
auto -> Ref

rniwa commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -397,8 +399,8 @@
         auto request = m_requestData->outstandingByteRangeRequests.take(identifier);
-        if (request.streamLoader())
-            cancelAndForgetStreamLoader(*request.streamLoader());
+        if (RefPtr streamLoader = request.protectedStreamLoader())
```
I think our convention is to call the regular non-protected version of `request.streamLoader` here and store it in RefPtr.

basuke replied:
Ah, I see. Still I do too much.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -320,7 +320,8 @@
     for (auto iterator = m_requestData->streamLoaderMap.begin(); iterator != m_requestData->streamLoaderMap.end(); iterator = m_requestData->streamLoaderMap.begin()) {
         removeOutstandingByteRangeRequest(iterator->value);
-        cancelAndForgetStreamLoader(*iterator->key);
+        Ref streamLoader = *iterator->key.get();
```
Doesn't need a local variable. Can use `Ref { *iterator->key.get() }` below.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
```diff
@@ -397,8 +398,8 @@
-        if (request.streamLoader())
-            cancelAndForgetStreamLoader(*request.streamLoader());
+        if (RefPtr streamLoader = request.streamLoader())
+            cancelAndForgetStreamLoader(*streamLoader.get());
```
no need for `.get()`

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -1095,24 +1095,25 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
-        row.contentsLayer->setAnchorPoint({ });
-        row.contentsLayer->setDrawsContent(true);
-        row.contentsLayer->setAcceleratesDrawing(m_plugin->canPaintSelectionIntoOwnedLayer());
+        RefPtr rowContentsLayer = row.contentsLayer;
```
What is this assignment for? You're reassigning the variable unconditionally on the next line..

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
```diff
@@ -1095,24 +1095,25 @@ static float elasticDeltaForTimeDelta(float initialPosition, float initialVeloci
-        row.selectionLayer->setAcceleratesDrawing(true);
-        row.selectionLayer->setBlendMode(BlendMode::Multiply);
-        m_layerIDToRowIndexMap.set(*row.selectionLayer->primaryLayerID(), rowIndex);
+        RefPtr rowSelectionLayer = row.selectionLayer;
```
Unnecessary assignment again?

basuke committed:
Use more smart pointers in PDF extension
https://bugs.webkit.org/show_bug.cgi?id=279472

Reviewed by Chris Dumez and Ryosuke Niwa.

Use smart pointer to protect arguments to be gone.

* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
(WebKit::ByteRangeRequest::protectedStreamLoader):
(WebKit::PDFIncrementalLoader::receivedNonLinearizedPDFSentinel):
(WebKit::PDFIncrementalLoader::incrementalPDFStreamDidReceiveData):
(WebKit::PDFIncrementalLoader::dataProviderGetBytesAtPosition):
(WebKit::PDFIncrementalLoader::dataProviderGetByteRanges):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::generatePreviewImageForPage):
(WebKit::AsyncPDFRenderer::serviceRequestQueue):
(WebKit::AsyncPDFRenderer::paintTilesForPage):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
(WebKit::PDFDiscretePresentationController::startTransitionAnimation):
(WebKit::PDFDiscretePresentationController::updateLayersForTransitionState):
(WebKit::PDFDiscretePresentationController::deviceOrPageScaleFactorChanged):
(WebKit::PDFDiscretePresentationController::buildRows):
(WebKit::PDFDiscretePresentationController::updateLayersOnLayoutChange):
(WebKit::PDFDiscretePresentationController::updateLayersAfterChangeInVisibleRow):
(WebKit::PDFDiscretePresentationController::repaintForIncrementalLoad):
(WebKit::PDFDiscretePresentationController::RowData::leftPageBackgroundLayer const):
(WebKit::PDFDiscretePresentationController::RowData::rightPageBackgroundLayer const):
(WebKit::PDFDiscretePresentationController::RowData::backgroundLayerForPageIndex const):

Canonical link: https://commits.webkit.org/283526@main
